pkg/apis/seedmanagement/v1alpha1/defaults.go

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"encoding/json"
+	"fmt"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	"github.com/gardener/gardener/pkg/apis/seedmanagement/helper"
+	configv1alpha1 "github.com/gardener/gardener/pkg/gardenlet/apis/config/v1alpha1"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func addDefaultingFuncs(scheme *runtime.Scheme) error {
+	return RegisterDefaults(scheme)
+}
+
+// SetDefaults_ManagedSeed sets default values for ManagedSeed objects.
+func SetDefaults_ManagedSeed(obj *ManagedSeed) {
+	switch {
+	case obj.Spec.SeedTemplate != nil:
+		// Set seed defaults
+		seed := gardencorev1beta1.Seed{Spec: obj.Spec.SeedTemplate.Spec}
+		gardencorev1beta1.SetDefaults_Seed(&seed)
+		obj.Spec.SeedTemplate.Spec = seed.Spec
+
+		// Set default seed secret references
+		setDefaultSecretRef(&obj.Spec.SeedTemplate.Spec, obj)
+		setDefaultBackupSecretRef(&obj.Spec.SeedTemplate.Spec, obj)
+
+	case obj.Spec.Gardenlet != nil && obj.Spec.Gardenlet.Config != nil:
+		// Initialize gardenlet config
+		var gardenletConfig *configv1alpha1.GardenletConfiguration
+		var err error
+		if obj.Spec.Gardenlet.Config != nil {
+			gardenletConfig, err = helper.DecodeGardenletConfigExternal(obj.Spec.Gardenlet.Config)
+			if err != nil {
+				return
+			}
+		} else {
+			gardenletConfig = &configv1alpha1.GardenletConfiguration{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: configv1alpha1.SchemeGroupVersion.String(),
+					Kind:       "GardenletConfiguration",
+				},
+			}
+		}
+
+		// Set gardenlet config defaults
+		configv1alpha1.SetDefaults_GardenletConfiguration(gardenletConfig)
```

Comment:
I don't agree. Without calling the defaults, the `GardenletConfiguration` validation will fail as well, unless the user specifies a number of things he shoudn't care specifying, such as server and SNI settings. And without validation, we let the user specify invalid values that will only blow up during the reconciliation. 

Besides, while testing I once commented this line out. To my surprise, the default values were still applied. I think it's done implicitly by the previous call `helper.DecodeGardenletConfigExternal(...)`. So I am not even sure how to disable the defaults.

Reply:
As I wrote above, to avoid the "merge with parent" issue, we decided to not default the `GardenletConfiguration` resource (since it breaks merging with parent), but to keep validation (since we do really need it). To enable this, I will change the validation code to "decouple" it from the defaults, e.g. making sure an empty `GardenletConfiguration` resource can pass validation.

I still need to find out how to actually achieve not setting the defaults during the decoding, but I hope there is a way.